util/EIP-712/src/lib.rs

```diff
@@ -0,0 +1,303 @@
+#[macro_use]
+extern crate serde_derive;
+extern crate serde_json;
+extern crate ethabi;
+extern crate ethereum_types;
+extern crate hex;
+extern crate keccak_hash;
+extern crate quickersort;
+extern crate failure;
+#[macro_use]
+extern crate failure_derive;
+
+mod eip712;
+mod error;
+
+pub use error::*;
+pub use eip712::*;
+
+use ethabi::{encode, Token};
+use ethereum_types::{Address, U256};
+use keccak_hash::{keccak, H256};
+use serde_json::Value;
+use std::collections::HashSet;
+use std::str::FromStr;
+
+/// given a type and HashMap<String, Vec<FieldType>>
+/// returns a HashSet of dependent types of the given type
+fn build_dependencies<'a>(message_type: &'a str, message_types: &'a MessageTypes) -> Option<(HashSet<&'a str>)>
+{
+	// get the associated FieldType of the given type
+	if let Some(fields) = message_types.get(message_type) {
+		let mut deps = HashSet::new();
+		deps.insert(message_type);
+
+		for field in fields {
+			// seen this type before? skip
+			if deps.contains(&*field.type_) {
+				continue;
+			}
+			if let Some(set) = build_dependencies(&field.type_, &message_types) {
+				deps.extend(set);
+			}
+		}
+		return Some(deps);
+	}
+	// primitive types like uint256, address wouldn't exist in MessageTypes
+	return None;
+}
+
+fn encode_type(message_type: &str, message_types: &MessageTypes) -> Result<String> {
+	let deps = {
+		if let Some(mut temp) = build_dependencies(message_type, message_types) {
+			temp.remove(message_type);
+			let mut temp = temp.into_iter().collect::<Vec<_>>();
+			quickersort::sort(&mut temp[..]);
+			temp.insert(0, message_type);
+			temp
+		} else {
+			return Err(ErrorKind::NonExistentType)?
+		}
+	};
+
+	let encoded = deps.into_iter().fold(String::new(), |mut acc, dep| {
+		// this unwrap is safe because we're asking for a dependency that was recently pulled out of message_types
+		let types = message_types.get(dep)
+			.unwrap()
+			.iter()
+			.map(|value| format!("{} {}", value.type_, value.name))
+			.collect::<Vec<_>>()
+			.join(",");
+
+		acc.push_str(&format!("{}({})", dep, types));
+		return acc;
+	});
+
+	Ok(encoded)
+}
+
+fn type_hash(message_type: &str, typed_data: &MessageTypes) -> Result<H256> {
+	Ok(keccak(encode_type(message_type, typed_data)?))
+}
+
+fn encode_data(message_type: &str, message_types: &MessageTypes, message: &Value) -> Result<Vec<u8>> {
+	let type_hash = (&type_hash(message_type, &message_types)?).to_vec();
+	let mut tokens = vec![Token::FixedBytes(type_hash)];
+	for field in message_types.get(message_type).ok_or_else(|| ErrorKind::NonExistentType)? {
+		let value = &message[&field.name];
+		match &*field.type_ {
+			// Array type e.g uint256[], string[]
+			ty if ty.rfind(']') == Some(ty.len() - 1) => {
+				let array_type = ty.split('[').collect::<Vec<_>>()[0];
+				let mut items = vec![];
+				for item in value.as_array().ok_or_else(|| serde_error("array", &field.name))? {
+					let encoded = encode_data(array_type.into(), &message_types, item)?;
```

Comment:
This is also recursive, might be good to actually limit the depth to prevent stack overflows.